bc/core/utils/images.py

```diff
@@ -0,0 +1,273 @@
+import io
+from dataclasses import dataclass, field
+from math import ceil, sqrt
+from pathlib import Path
+from textwrap import fill, wrap
+
+from PIL import Image, ImageFont
+from PIL.ImageDraw import Draw
+
+from bc.settings import STATIC_GLOBAL_ROOT
+
+
+@dataclass
+class TextImage:
+    title: str
+    description: str
+    title_font_path: Path = (
+        STATIC_GLOBAL_ROOT / "fonts" / "CooperHewitt-Bold.otf"
+    )
+    desc_font_path: Path = (
+        STATIC_GLOBAL_ROOT / "fonts" / "CooperHewitt-Light.otf"
+    )
+    font_size: int = 24
+    font_spacing: int = 16
+    padding: float = 10.0
+    format: str = "png"
+    line_height: int = field(init=False)
+    width: int = field(init=False)
+    height: int = field(init=False)
+
+    def __post_init__(self):
+        self.title_font = ImageFont.truetype(
+            str(self.title_font_path), size=self.font_size
+        )
+        self.desc_font = ImageFont.truetype(
+            str(self.desc_font_path), size=self.font_size
+        )
+        self.line_height = self.font_size + self.font_spacing
+
+    def get_text_length(self, str: str) -> int:
+        """
+        Returns the length(in pixels) of the text when rendered. This methods assumes
+        the text will be rendered in a single line and includes extra margins for some
+        fonts.
+
+        Args:
+            str (str): Text to render.
+
+        Returns:
+            int: Length of given text.
+        """
+        bounding_box = self.desc_font.getbbox(str)
+        return bounding_box[2]
+
+    def get_available_space(self, wrapped: list[str]) -> int:
+        """
+        Returns the number of pixels available in the horizontal axis of the rectangle.
+        This method uses the longest line from the given array and subtracts its length
+        (in pixels) from the width of the rectangle.
+
+        Args:
+            wrapped (list[str]): list of lines without final newlines.
+
+        Returns:
+            int: number of available pixels in the horizontal axis
+        """
+        return self.width - self.get_text_length(max(wrapped, key=len))
+
+    def get_height_approximation(self, length: int) -> int:
+        """
+        Returns a rough approximation of the minimum height required to render the given
+        number of pixels in a 16:9 image. This method does not include padding in the
+        equation to keep it simple.
+
+        Args:
+            length (int): Number of the pixels of the text when rendered.
+
+        Returns:
+            int: height (in pixels)
+        """
+        return ceil(sqrt((9 / 16) * (self.line_height) * length))
```

Comment:
Well, I might be able to make this make sense if I stare at it awhile longer, but is there something that you can do to make it intuitive why this works? 

Also, please remove extra parens:

```suggestion
        return ceil(sqrt((9 / 16) * self.line_height * length))
```

Reply:
@mlissner let's suppose We have a **rectangle** with height **H** and width **W** and this **rectangle** has a **16:9** aspect ratio (Mastodon crops the images using that ratio), then the following  is always `true`:

``` 
 1) W/H = 16/9
```

If We want to render a paragraph inside this **rectangle**, We need to follow these rules:

``` 
 a) the width of each line should be W pixels at most ( this avoids horizontal overflow)
 b) the height of the paragraph (after wrapping the text) should be less than H ( This avoids vertical overflow)
```

We can write **b** like the following inequation:

```
 2) H ≥ paragraph_height 
```

We can approximate the value of the **paragraph_height** using the **number of lines** and the **line height**, so:

```
3) paragraph_height ≈ line_count * line_height
```
Now, We can use the expression from (3) in (2). Hence, inequation (2) becomes

```
4) H ≥ line_count * line_height
```

We still need the **line_count** but We can compute the width (in pixels) of the text without wrapping (the text is rendered as a single line) and then We can use rule **a** to get an estimate of the number of lines.  Hence, We can express **line_count** like:

```
5) line_count ≈ ( width_single_line / W )
```

If we use (5) in inequation (4), We got the following:

```
6)  H ≥ ( width_single_line / W ) * line_height 

7)  H ≥ ( width_single_line * line_height)/ W 
```

We can use (1) and solve for **W**. Hence, equation (7) becomes:

```
8) H ≥ ( width_single_line * line_height)/ [ (16/9) * H ]

9) H ≥ ( width_single_line * line_height ) / ( 16 / 9)

10) H ≥ √ [ (9/16) * ( width_single_line * line_height) ]
```

We're using inequation 10 to get the initial dimensions of the rectangle ( We didn't include padding in the equation to keep it simple). 

Let me know what you think.

Reply:
Goodness, this was quite a puzzle. I think we either need this explanation in the code or a reference to it. My hope is that future tweaks to the code won't have to figure this out all over again.